Azure container registry build feature GA API specification and examples.

mnltejaswini opened the pull request:
This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process.

### PR information
- [x] The title of the PR is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [x] If applicable, the PR references the bug/issue that it fixes.
- [ ] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [ ] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [ ] My spec meets the review criteria:
  - [ ] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [ ] The spec follows the guidelines described in the [Swagger checklist](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md) document.
  - [ ] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

mnltejaswini committed:
Azure container registry build GA API specification and examples.

azuresdkci commented:
Can one of the admins verify this patch?

AutorestCI commented:
# Automation for azure-sdk-for-go

AutorestCI commented:
# Automation for azure-sdk-for-java

annatisch commented:
Adding @ravbhatnagar to review new ARM API version swagger.

@mnltejaswini - there are some model validation errors in the examples - could you take a look?https://travis-ci.org/Azure/azure-rest-api-specs/jobs/414189551

Thanks!

mnltejaswini committed:
Fix the Model validation errors in example files.

mnltejaswini committed:
Removed secrets from Task properties as this is not yet supported.

annatisch commented:
Thanks for the updates @mnltejaswini 
Could you please update the readme tag so that the generated SDKs pick up the latest API version?
https://github.com/Azure/azure-rest-api-specs/pull/3617/files#diff-4837e14bb0f77202f522d1f223f87c3aR29

Unless you have a particular reason not to publish SDKs for this API?

Finally, there's still a couple of errors in the Tasks_Get example:
```
 error: 
operationId: Tasks_Get
scenario: Tasks_Get
source: response
responseCode: '200'
severity: 0
errorCode: OBJECT_MISSING_REQUIRED_PROPERTY
errorDetails:
  code: OBJECT_MISSING_REQUIRED_PROPERTY
  params:
    - name
  message: 'Missing required property: name'
  path: properties/trigger/sourceTriggers/0
  title: SourceTrigger
  description: The properties of a source based trigger.

 error: 
operationId: Tasks_Get
scenario: Tasks_Get
source: response
responseCode: '200'
severity: 0
errorCode: OBJECT_MISSING_REQUIRED_PROPERTY
errorDetails:
  code: OBJECT_MISSING_REQUIRED_PROPERTY
  params:
    - name
  message: 'Missing required property: name'
  path: properties/trigger/baseImageTrigger
  title: BaseImageTrigger
  description: The trigger based on base image dependency.
```

mnltejaswini committed:
Removed Alias and secrets  property from task.

johanste commented on specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/stable/2018-09-01/containerregistry_build.json:
```diff
@@ -0,0 +1,1807 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.ContainerRegistry/registries/{registryName}/getBuildSourceUploadUrl": {
```
Is this API ever expected to be consumed from an ARM template? If so, @ravbhatnagar, is it still true that the name should be list*** (regardless of how many items are actually returned)?

mnltejaswini replied:
Yes, you are correct. It cannot be called from ARM template if it is not prefixed with list. There is no use case for this API to be called from template

johanste replied:
In general, I've often found that there are scenarios for consuming pretty much any computed value from within a template. So, my default recommendation is to take an "assume it will be consumed from within a template unless proven otherwise" approach. But this is a general statement. If you truly don't have any scenarios where one would, say, provision a virtual machine that needs the access the value (or that user assigned identities would enable the access from the resource itself), then this is all good.

mnltejaswini replied:
I renamed it to prefix with list so that it can be invoked from template.

mnltejaswini committed:
Fixed the readme to the latest tag.

johanste commented on specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/stable/2018-09-01/containerregistry_build.json:
```diff
@@ -0,0 +1,1807 @@
+          {
+            "name": "$filter",
+            "in": "query",
+            "description": "The runs filter to apply on the operation.",
```
What capabilities does the filter have? Full odata (operators and all properties)? If not, specifying what the user can put in the $filter in the description is very helpful...

mnltejaswini replied:
Makes sense. there are some restrictions on the odata query , I will add them to the description.

mnltejaswini replied:
I updated the description to include restrictions on filter.

johanste commented on specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/stable/2018-09-01/containerregistry_build.json:
```diff
@@ -0,0 +1,1807 @@
+            "format": "int32"
+          },
+          {
+            "name": "$skipToken",
```
How would the user get the skip token value? If the only way to get it is from a nextLink in the response, then there is limited value in documenting the parameter, since the only way that the user can get it is by parsing the URL, and then putting it back into the API to build up exactly the same URL as they got in the nextLink to begin with....

mnltejaswini replied:
The nextLink is self contained and can be invoked as is. Please correct me if my understanding is wrong?

johanste replied:
Is there some other way that I, as a user, can determine what value I could possibly pass in for the $skipToken parameter?

mnltejaswini replied:
No, the only way is from nextLink

mnltejaswini replied:
Got it

mnltejaswini commented:
@johanste  Is it allowed for the PATCH and PUT request payload is of same type? The lint checker complaints about the conflicts between required and patchable properties. So we either remove Required for Task and do only server validation versus separate the create and update parameters. Which one is preferred? Can you please advise

johanste commented:
@mnltejaswini, having required parameters for a PATCH request body rarely makes sense. You want to provide only the value that you want to modify, which is orthogonal to if they are required on create/PUT. 

The most common pattern is to have separate models. Unfortunately, there is no way in swagger/openapi to have selectively required parameters depending on the point of usage.

mnltejaswini committed:
separate the PUT and PATCH request payload

mnltejaswini commented:
@johanste  Thank you. Will separate the properties.

mnltejaswini committed:
Matching the resource schema with that in the Registry swagger

ravbhatnagar commented:
pending a couple of updates from @mnltejaswini. Please let me know once you are ready and I can review.

mnltejaswini committed:
added more description to $filer, separated create and update parameters to avoid conflict between required nd patchable properties.

mnltejaswini commented:
@ravbhatnagar  I separated the create and update parameters to avoid conflict between required and patchable properties. Can you please review.

mnltejaswini committed:
Replaced GitCOmmitTrigger to SourceTrigger

mnltejaswini commented:
@ravbhatnagar  Can you please review

ravbhatnagar commented:
This was reviewed in person. Signing off!

annatisch commented:
Thanks @ravbhatnagar!
@mnltejaswini - could you please take a look at the validation errors in the examples?

mnltejaswini committed:
Added isArchiveEnabled as part of Schedule Run request,  renamed the arguments property to values for Task type requests and steps

mnltejaswini commented:
@ravbhatnagar  As per the review board feedback, there are minor changes, added isArchiveEnabled as part of schedule run request and renamed the arguments to values property for task type step and run requests.

mnltejaswini commented:
@annatisch  where can I find the validation errors in examples?

annatisch commented:
Thanks @mnltejaswini - you can find them here:
https://travis-ci.org/Azure/azure-rest-api-specs/jobs/419797791

I will also review the PR today.

mnltejaswini commented:
@annatisch  They doesn't seem to be valid. The payload for GET and PUT is the same, while we can PUT some sensitive information as part of resource creation, we cannot return them as part of GET. This has be the pattern we were using for all our resources.

mnltejaswini committed:
Removed the Required attribute for the source control auth properties to fix the model validation failures.

mnltejaswini committed:
Update readme.md to include tags for all languages

annatisch reviewed (changes requested):
Just a few small things :)

mnltejaswini commented on specification/containerregistry/resource-manager/Microsoft.ContainerRegistry/stable/2018-09-01/containerregistry_build.json:
```diff
@@ -0,0 +1,2247 @@
+          "Tasks"
+        ],
+        "description": "Returns a task with extended information that includes all secrets.",
+        "operationId": "Tasks_ListDetails",
```
I see, ok then I will change the operation ID

mnltejaswini committed:
Updated the operation ID as per the feedback
